service/popm/popm_test.go

```diff
@@ -12,6 +12,7 @@ import (
 	"fmt"
```

Comment:
While I cannot put my finger on it, it seems not unlikely that there is a logic race/error somewhere in this code. I would create a test that has hundreds of go routines attacking the pipeline and see what `test -race` comes up with.